### PR TITLE
L65: Additional PyPI Packages for gRPC Python

### PR DESCRIPTION
**Pull request by lidizheng**
> gRPC Python is uploaded as `grpcio` on PyPI, but there is another package named
`grpc`. Hence some users are confused. This document proposes gRPC Python to
upload additional packages named `grpc` and `grpc-*` as proxies to our official
packages.

### CODE REVIEW

File: L65-python-package-name.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+### Package `grpcio`
+
+Package `grpcio` is first uploaded in 2015, given the fact that the name `grpc`
+already being taken, this name is likely a workaround.
+
+## Proposal
+
+### List of Package Name
+
+Here is a list of new proxy packages:
```
Comment (gnossen) [inline L65-python-package-name.md:35]: I"d recommend you move the section defining "proxy package" to before this first usage of the term.

File: L65-python-package-name.md [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+* `grpc`
+* `grpc-status`
+* `grpc-channelz`
+* `grpc-tools`
+* `grpc-reflection`
+* `grpc-testing`
+* `grpc-health-checking`
+
+### Proxy Mechanism
```
Comment (gnossen) [inline L65-python-package-name.md:44]: "Proxy" isn't the right name for what we're doing here. What we're really doing is nerfing the `grpc` package with a message suggesting that they install `grpcio` instead. This mechanism is the sign in this picture: 

Whereas a proxy would actually be a cop.

File: L65-python-package-name.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+```
+
+### Maintenance Cost
+
+The maintenance cost for this solution is minimum since these packages are
+agnostic to our releases.
+
+### Package Version
+
+All new proxy packages will be versioned as `1.0`.
```
Comment (gnossen) [inline L65-python-package-name.md:61]: I didn't fully appreciate what this meant the first time I read through. I would reword to something more along the lines of "Rather than releasing new versions of the `grpc-*` packages in lockstep with the canonical `grpcio-*` packages for each release, we will instead upload only one new artifact for each package name, versioned as '1.0'.". You might also want to make an argument for this strategy. If we go with this strategy as opposed to uploading a new copy of the same `RunTimeError`-producing package, then people running `pip install grpc==1.29.0` will get a "package not found" error instead of a more informative "I think you have the wrong number"-style message.

File: L65-python-package-name.md [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+gRPC Python is uploaded as `grpcio` on PyPI, but there is another package named
+`grpc`. Hence some users are confused. This document proposes gRPC Python to
+upload additional packages named `grpc` and `grpc-*` as proxies to our official
+packages.
+
+## Background
+
+### Package `grpc`
+
+Package `grpc` is an alternative version of gRPC Python only supports Python 2,
```
Comment (gnossen) [inline L65-python-package-name.md:21]: We probably want to add a bit of history as well on how we obtained ownership of the package name.

Comment (lidizheng) [reply L65-python-package-name.md:21]: Add a link to https://github.com/pypa/pypi-support/issues/3.

File: L65-python-package-name.md (hunk 1 above)
Comment (lidizheng) [reply L65-python-package-name.md]: Good idea. Moved.

File: L65-python-package-name.md (hunk 2 above)
Comment (lidizheng) [reply L65-python-package-name.md:34]: How should we call it? "Cop packages" / "Guard packages" / "Watchdog packages" / "Toxic packages"? What do you recommend?

File: L65-python-package-name.md (hunk 3 above)
Comment (lidizheng) [reply L65-python-package-name.md]: Thanks for the suggestion. Updated to:
```
Rather than releasing new versions of the `grpc` and `grpc-*` packages in
lockstep with the canonical `grpc` and `grpcio-*` packages for each release, we
will instead upload only one new artifact for each package name, versioned as
`v1.0`. A stable version number suggests this package's behavior won't change
actively.
```

I reword this section, so it is less confusing. For people who want to pin package version, they are using gRPC Python in a more serious scenario. I think they would need to double check the package name.

I'm also okay to upload all future versions with a script. But they would taken our ability to change its logic forever. Instrumenting our infrastructure to produce more artifacts for each release is time-costly, we might not want to make existing release procedure more complex.

Also, separated version suggests the nature of release agnostic for the new packages.

File: L65-python-package-name.md (hunk 2 above)
Comment (gnossen) [reply L65-python-package-name.md:34]: No. I didn't mean things quite so literally with the meme. "Archived package" or "retired package" is probably best.

File: L65-python-package-name.md (hunk 3 above)
Comment (gnossen) [reply L65-python-package-name.md]: > lockstep with the canonical `grpc` and `grpcio-*`

The current wording implies that both `grpc` and `grpcio` prefixes are canonical. How about `grpcio*`? That includes `grpcio`, `grpcio-channelz`, etc.

> I'm also okay to upload all future versions with a script. But they would taken our ability to change its logic forever

I don't think that's true. We can always decide that the next release of `grpc` will contain something other than a `RunTimeError`.

> Instrumenting our infrastructure to produce more artifacts for each release is time-costly, we might not want to make existing release procedure more complex.

Totally agree.

File: L65-python-package-name.md (hunk 2 above)
Comment (lidizheng) [reply L65-python-package-name.md]: It's sad that those new packages never shined, but went straight to retirement. Perhaps that's been the story of life. Could it be better to reference it as just "package", and avoid using the word "proxy"?

Comment (gnossen) [reply L65-python-package-name.md]: They did have a chance to shine. As "gevent RPC". *That's* the package that I'm referring to as archived, not a hypothetical "gRPC" `grpc` package.

I'm fine referring to them as just "the `grpc`-prefixed packages" or something like that. So long as it's clear in each instance whether we're referring to `grpcio`, `grpcio-reflection`, etc. *or* `grpc`, `grpc-testing`, etc.
